contract_payment_mode/models/contract.py

```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+from openerp import models, fields, api
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = 'account.analytic.account'
+
+    payment_mode_id = fields.Many2one(
+        'payment.mode',
+        string='Payment Mode',
+        domain="[('type', '=', 'sale')]")
+
+    @api.multi
+    def on_change_partner_id(self, partner_id, name):
+        res = super(AccountAnalyticAccount, self).on_change_partner_id(
```

Comment:
For the records, all API on_changes (triggered by UI), although decorated with @api.multi should be multi-record safe, because you mustn't use self (or make a browse in old API) to get current values, that are only loaded on client-side. That's why you pass extra arguments with the values you need to work with (and the problem that the module _web_context_tunnel_ solves for adding more arguments).